contrib/pg_tde/src/bin/pg_tde_archive_decrypt.c

```diff
@@ -0,0 +1,228 @@
+#include "postgres_fe.h"
+
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "access/xlog_internal.h"
+#include "access/xlog_smgr.h"
+#include "common/logging.h"
+
+#include "access/pg_tde_fe_init.h"
+#include "access/pg_tde_xlog_smgr.h"
+
+static bool
+is_segment(const char *filename)
+{
+	return strspn(filename, "0123456789ABCDEF") == 24 && filename[24] == '\0';
+}
+
+static void
+write_decrypted_segment(const char *segpath, const char *segname, int pipewr)
+{
+	int			fd;
+	off_t		fsize;
+	int			r;
+	int			w;
+	TimeLineID	tli;
+	XLogSegNo	segno;
+	PGAlignedXLogBlock buf;
+	off_t		pos = 0;
+
+	fd = open(segpath, O_RDONLY | PG_BINARY, 0);
+	if (fd < 0)
+		pg_fatal("could not open file \"%s\": %m", segname);
+
+	/*
+	 * WalSegSz extracted from the first page header but it might be
+	 * encrypted. But we need to know the segment seize to decrypt it (it's
+	 * required for encryption offset calculations). So we get the segment
+	 * size from the file's actual size. XLogLongPageHeaderData->xlp_seg_size
+	 * there is "just as a cross-check" anyway.
+	 */
+	fsize = lseek(fd, 0, SEEK_END);
+	XLogFromFileName(segname, &tli, &segno, fsize);
+
+	r = xlog_smgr->seg_read(fd, buf.data, XLOG_BLCKSZ, pos, tli, segno, fsize);
+
+	if (r == XLOG_BLCKSZ)
+	{
+		XLogLongPageHeader longhdr = (XLogLongPageHeader) buf.data;
+		int			walsegsz = longhdr->xlp_seg_size;
+
+		if (walsegsz != fsize)
+			pg_fatal("mismatch of segment size in WAL file \"%s\" (header: %d bytes, file size: %ld bytes)",
+					 segname, walsegsz, fsize);
+
+		if (!IsValidWalSegSize(walsegsz))
```

Comment:
Couldn't this part be a helper function to avoid the duplication in the two tools? And `is_segment` is also common

Reply:
I don't think the duplicated code is enough to bother but I could give it a shot and see if anything improves.